File: internal/vision/vision.go
// Package vision wraps the local Ollama HTTP API to ask SmolVLM2 yes/no
// questions about screenshots, plus narrate and summarize runs.
package vision

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	functionalPrompt = "You are a QA test assistant. Look at the screenshot of a web application. " +
		"Answer the following question with 'YES' or 'NO' as the FIRST word, then a " +
		"one-sentence reason. Question: %s"
	narratePrompt = "In ONE short sentence, describe what is visible on this web app screenshot. " +
		"Mention specific UI elements (buttons, forms, errors, content). No preamble."
	summaryPrompt = "You are a senior QA engineer. Below is a per-step log from one regression " +
		"run. Write 2-3 sentences summarizing what was tested, what worked, and what failed. " +
		"Be specific. No bullets, no markdown.\n\nRun log:\n%s"
)

type Client struct {
	host   string
	model  string
	client *http.Client
}

func NewClient(host, model string, timeout time.Duration) *Client {
	return &Client{
		host:   strings.TrimRight(host, "/"),
		model:  model,
		client: &http.Client{Timeout: timeout},
	}
}

// chatRequest is the subset of the Ollama /api/chat payload we use.
type chatRequest struct {
	Model    string                   `json:"model"`
	Messages []map[string]interface{} `json:"messages"`
	Options  map[string]interface{}   `json:"options"`
	Stream   bool                     `json:"stream"`
}

type chatResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
}

// ask sends a single-message chat call with optional images.
func (c *Client) ask(prompt string, imagePaths []string) (string, error) {
	images := make([]string, 0, len(imagePaths))
	for _, p := range imagePaths {
		b, err := os.ReadFile(p)
		if err != nil {
			return "", fmt.Errorf("read %s: %w", p, err)
		}
		images = append(images, base64.StdEncoding.EncodeToString(b))
	}
	body := chatRequest{
		Model: c.model,
		Messages: []map[string]interface{}{
			{"role": "user", "content": prompt, "images": images},
		},
		Options: map[string]interface{}{
			"temperature": 0.0,
			"num_predict": 200,
		},
		Stream: false,
	}
	buf, _ := json.Marshal(body)
	resp, err := c.client.Post(c.host+"/api/chat", "application/json", bytes.NewReader(buf))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return "", fmt.Errorf("ollama returned %d", resp.StatusCode)
	}
	var out chatResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return "", err
	}
	return strings.TrimSpace(out.Message.Content), nil
}

// Ping returns true if Ollama is reachable.
func (c *Client) Ping() bool {
	req, _ := http.NewRequest("GET", c.host+"/api/version", nil)
	resp, err := c.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode == 200
}

// Verdict is the parsed outcome of a yes/no AI check.
type Verdict struct {
	Passed bool
	Raw    string
}

// FunctionalCheck asks the model a yes/no question about a screenshot.
func (c *Client) FunctionalCheck(imagePath, question string, expectYes bool) (Verdict, error) {
	raw, err := c.ask(fmt.Sprintf(functionalPrompt, question), []string{imagePath})
	if err != nil {
		return Verdict{Raw: err.Error()}, err
	}
	first := strings.ToUpper(strings.TrimRight(strings.Fields(raw+" ")[0], ".,:;!?"))
	gotYes := strings.HasPrefix(first, "YES")
	return Verdict{Passed: gotYes == expectYes, Raw: raw}, nil
}

// Narrate returns one short sentence describing what's on screen.
func (c *Client) Narrate(imagePath string) (string, error) {
	raw, err := c.ask(narratePrompt, []string{imagePath})
	if err != nil {
		return "", err
	}
	// Trim to one sentence if the model rambled.
	if idx := strings.IndexAny(raw, "\n"); idx > 0 {
		raw = raw[:idx]
	}
	return strings.TrimSpace(raw), nil
}

// SummarizeRun produces a 2-3 sentence verdict for an entire run log.
func (c *Client) SummarizeRun(runLog string) (string, error) {
	if len(runLog) > 3000 {
		runLog = runLog[len(runLog)-3000:]
	}
	return c.ask(fmt.Sprintf(summaryPrompt, runLog), nil)
}

File: src/moonlight/vision.py
"""SmolVLM2 client — wraps the local Ollama API for QA-flavoured prompts."""
from __future__ import annotations

import base64
from dataclasses import dataclass
from pathlib import Path

import ollama


FUNCTIONAL_PROMPT = (
    "You are a QA test assistant. Look at the screenshot of a web application. "
    "Answer the following question with 'YES' or 'NO' as the FIRST word, then a "
    "one-sentence reason. Question: {question}"
)

OCR_PROMPT = (
    "Transcribe all visible text in this screenshot exactly as shown. "
    "Output the text only, no commentary, no markdown."
)

REGRESSION_PROMPT = (
    "Two screenshots of the same web page are provided — the FIRST is the baseline "
    "and the SECOND is the current run. First, answer 'MEANINGFUL' or 'COSMETIC' "
    "as the FIRST word: MEANINGFUL = a real UI/bug change, COSMETIC = noise such "
    "as timestamps, ads, random IDs, or animations. Then write ONE short sentence "
    "describing exactly what visibly changed between the two screenshots."
)

NARRATE_PROMPT = (
    "In ONE short sentence, describe what is visible on this web app screenshot. "
    "Mention specific UI elements (buttons, forms, errors, content). No preamble, "
    "no 'this image shows' — just the description."
)

SUMMARY_PROMPT = (
    "You are a senior QA engineer. Below is a per-step log from one regression "
    "run. Write 2–3 sentences summarizing: (1) what was tested, (2) what worked, "
    "(3) what failed and the likely cause. Be specific and concrete. No bullet "
    "points, no markdown. Output the summary only.\n\n"
    "Run log:\n{log}"
)


@dataclass
class VisionVerdict:
    passed: bool
    raw: str

    @property
    def reason(self) -> str:
        parts = self.raw.split(maxsplit=1)
        return parts[1].strip() if len(parts) > 1 else ""


class SmolVLM2Client:
    def __init__(
        self,
        host: str = "http://localhost:11434",
        model: str = "ahmadwaqar/smolvlm2-2.2b-instruct",
        timeout_s: int = 120,
        options: dict | None = None,
    ) -> None:
        self.model = model
        self._client = ollama.Client(host=host, timeout=timeout_s)
        self._options = {"temperature": 0.0, "num_predict": 200, **(options or {})}

    def _ask(self, prompt: str, images: list[Path]) -> str:
        encoded = [base64.b64encode(p.read_bytes()).decode("ascii") for p in images]
        resp = self._client.chat(
            model=self.model,
            options=self._options,
            messages=[{"role": "user", "content": prompt, "images": encoded}],
        )
        return (resp.get("message") or {}).get("content", "").strip()

    def functional_check(self, image: Path, question: str, expect_yes: bool = True) -> VisionVerdict:
        raw = self._ask(FUNCTIONAL_PROMPT.format(question=question), [image])
        first = raw.split(maxsplit=1)[0].strip(".,:;!?").upper() if raw else ""
        got_yes = first.startswith("YES")
        return VisionVerdict(passed=(got_yes == expect_yes), raw=raw)

    def ocr(self, image: Path) -> str:
        return self._ask(OCR_PROMPT, [image])

    def regression_judge(self, baseline: Path, current: Path) -> VisionVerdict:
        raw = self._ask(REGRESSION_PROMPT, [baseline, current])
        first = raw.split(maxsplit=1)[0].strip(".,:;!?").upper() if raw else ""
        # passed = differences are cosmetic (not real regressions)
        return VisionVerdict(passed=first.startswith("COSMETIC"), raw=raw)

    def narrate(self, image: Path) -> str:
        """Return one short sentence describing the screen — for the report."""
        raw = self._ask(NARRATE_PROMPT, [image])
        # Trim to one sentence if the model rambled
        for sep in ("\n", ". "):
            if sep in raw:
                raw = raw.split(sep, 1)[0].rstrip(".") + "."
                break
        return raw.strip()

    def summarize_run(self, run_log: str) -> str:
        """2–3 sentence verdict over an entire run. Pure text — no images."""
        # Trim to ~3000 chars so the model context isn't blown.
        log = run_log[-3000:]
        try:
            resp = self._client.chat(
                model=self.model,
                options={**self._options, "num_predict": 300},
                messages=[{"role": "user", "content": SUMMARY_PROMPT.format(log=log)}],
            )
            return (resp.get("message") or {}).get("content", "").strip()
        except Exception as e:
            return f"(AI summary unavailable: {e})"

    def ping(self) -> bool:
        try:
            self._client.list()
            return True
        except Exception:
            return False
